### ISSUE TITLE
stuck in a loop

### ISSUE THREAD

**Issue by kishwarshafin** [nanoporetech/medaka#59]
Hi, I have running medaka on a GPU machine, but it looks like the call consensus gets stuck in a loop after a while.

[23:00:04 - PWorker] All done, 0 remainder regions.
[23:00:04 - PWorker] Running inference for 0.0M draft bases.
[23:00:04 - Sampler] Initializing sampler for consensus of region 4684:48-3642.
[23:00:04 - Feature] Processed 4684:48.0-3641.0 (median depth 4.0)
[23:00:04 - Sampler] Took 0.13s to make features.
[23:00:04 - Sampler] Pileup for 4684:48.0-3641.0 is of width 4324
[23:02:19 - PWorker] All done, 0 remainder regions.
[23:02:19 - PWorker] Running inference for 0.0M draft bases.
[23:02:19 - Sampler] Initializing sampler for consensus of region 6506:749-8932.
[23:02:19 - Feature] Processed 6506:749.0-8931.0 (median depth 1.0)
[23:02:19 - Sampler] Took 0.13s to make features.
[23:02:19 - Sampler] Pileup for 6506:749.0-8931.0 is of width 8875
[23:04:34 - PWorker] All done, 0 remainder regions.
[23:04:34 - PWorker] Running inference for 0.0M draft bases.
[23:04:34 - Sampler] Initializing sampler for consensus of region 2556:0-7245.
[23:04:34 - Feature] Processed 2556:0.0-7244.0 (median depth 3.0)
[23:04:34 - Sampler] Took 0.12s to make features.
[23:04:34 - Sampler] Pileup for 2556:0.0-7244.0 is of width 8008

Also, you guys reported that polishing human genome takes ~5h. I have tried almost all possible parameters to make things finish under 10h, but the stitch itself takes more than 5 hours. Am I missing something here?

**Comment by cjw85** [comment]
This looks like the expected output during the phase in which medaka is processing short contigs: contigs where the RNN input matrix has a shorter time dimension than the default width. No GPU batching is implemented for this stage as we haven't yet found a good method which balances performance and accuracy. This is a point of the code we are continually revisiting.

For small and contiguous assemblies the current implementation is not usually egregiously slow; judging by the numbers (region lengths and depths) above it looks like this is not the case for your assembler?

The inference performance of medaka is roughly 1 minute per 10 Mbases on a V100 GPU with large contigs. Under these conditions the bottleneck is currently parsing cigar strings in htslib (`resolve_cigar2`)!

**Comment by kishwarshafin** [comment]
Ok, I see. I went through the codebase of Medaka today. To my suggestion, you have to consider "data generation" and "inference" separately to take full advantage of modern machine learning libraries. We started with a minimized logit version like medaka to assist the new Shasta assembler with polishing which is [PEPPER](https://github.com/kishwarshafin/pepper) (this is not finished, sorry for the poor documentation), if you look into the implementation you'll see that we generate the data first so we can take advantage of mini-batch offered by PyTorch/TensorFlow if you do it like that then the inference time has no "data generation" halt. This online method of generating data and inferring at the same time is intriguing but puts a massive burden in run-time.

PEPPER uses pybind to expose htslib library to python [HERE](https://github.com/kishwarshafin/pepper/blob/master/modules/headers/dataio/bam_handler.h). Hence we generate all the images under 2 hours. And the inference is about 2 hours on a GPU with a batch size of 1024. I hope you see the point I'm making here. Also, [HELEN](https://github.com/kishwarshafin/helen) uses almost the same mechanism, and the inference time is 3 hours, and stitching is under an hour. 

We have struggled with Medaka's run-time while comparing with our tools. We are data limited; hence, we had to do things intelligently so we can make a model with minimum data-points. You have all the data in the world; I think if you take care of these things, Medaka will be a great tool to use.

Thanks for clarifying the query, I'll close the issue. Cheers.
